sphinx_immaterial/kbd_keys.py

```diff
@@ -0,0 +1,112 @@
+from typing import List, Tuple
+from docutils import nodes
+from sphinx.application import Sphinx
+from sphinx.config import Config
+from sphinx.writers.html import HTMLTranslator
+from sphinx.util.logging import getLogger
+
+LOGGER = getLogger(__name__)
+
+try:
+    import pymdownx.keymap_db as keys_db
+except ImportError:
+    LOGGER.info(
+        "Could not import `keymap_db` module from `pymdownx` package.\n    "
+        "Please ensure `pymdown-extensions` is installed.\n    "
+        "The `:keys:` role has no default key map."
+    )
+    keys_db = None
+
+
+class spanNode(nodes.container):
+    pass
+
+
+def visit_span(self: HTMLTranslator, node: spanNode):
+    self.body.append("<span")
+    if "classes" in node and node.get("classes"):
+        self.body.append(
+            f' class="{" ".join([self.attval(cls) for cls in node["classes"]])}"'
+        )
+    if "text" in node:
+        self.body.append(">" + node["text"])
+    else:
+        self.body.append(">")
+
+
+def depart_span(self: HTMLTranslator, node: spanNode):
+    self.body.append("</span>")
+
+
+class kbdNode(nodes.container):
+    pass
+
+
+def visit_kbd(self, node):
+    self.body.append(
+        '<kbd class="' + f'{" ".join([self.attval(cls) for cls in node["classes"]])}'
+    )
+    if "text" in node:
+        self.body.append('">' + node["text"])
+    else:
+        self.body.append('">')
+
+
+def depart_kbd(self, node):
+    self.body.append("</kbd>")
+
+
+def keys_role(
+    role, rawtext, text, lineno, inliner, options={}, content=[]
+) -> Tuple[List[nodes.Node], List[str]]:
+    keys = text.split(KEYS_OPTS["keys_sep"])
+    keys_div = spanNode("", classes=[KEYS_OPTS["keys_class"]])
+    if KEYS_OPTS["keys_strict"]:
+        keys_div = kbdNode("", classes=[KEYS_OPTS["keys_class"]])
+    keys_out = []
+
+    def map_filter(key: str) -> Tuple[str, str]:
+        display = key.title()
+        cls = key.replace("_", "-").replace(" ", "-").lower()
+        if key in KEYS_OPTS["keys_map"].keys():
+            display = KEYS_OPTS["keys_map"][key]
+        if keys_db is not None:
+            if key in keys_db.aliases:
+                display = keys_db.keymap[keys_db.aliases[key]]
+                cls = keys_db.aliases[key]
+        return (cls, display)
+
+    for i, key in enumerate(keys):
+        key_cls, key_display = map_filter(key.strip().lower())
+        keys_key = kbdNode("", classes=["key-" + key_cls])
+        keys_key["text"] = key_display
+        keys_out.append(keys_key)
+        if i + 1 != len(keys):
+            keys_sep = spanNode("", classes=[])
+            keys_sep["text"] = KEYS_OPTS["keys_sep"]
+            keys_out.append(keys_sep)
+    keys_div += keys_out
+    return [keys_div], []
+
+
+KEYS_OPTS = {"keys_map": ({} if keys_db is None else keys_db.keymap.copy())}
```

Comment:
From the visit_kbd method you can access `self.builder.config` --- that allows you to avoid needing to use a global variable for `KEYS_OPT`.

Reply:
Thank you I had a hell of a time trying to figure out how to do that. I will give that a try.

Reply:
I remember why I resorted to a global var. The parsing of the interpreted text (`given_text.split(separator)`) is done from the role function. I couldn't find a way to use the config options from the role function Is there an role-function arg that provides access to the config vals?

The parsing should be done at the role function because it creates the necessary nodes. Maybe this can be done from the `visit_*()`, but I haven't tried it.

Reply:
ok its super hard to differentiate segments of a `keys` interpreted text at the `visit_kbd()` context. You should be aware of the HTML output for this role:

```rst
:keys:`ctrl+alt+tab`
```
results in
```html
<span class="keys">
  <kbd class="key-control">Ctrl</kbd>
  <span>+</span>
  <kbd class="key-alt">Alt</kbd>
  <span>+</span>
  <kbd class="key-tab">Tab</kbd>
</span>
```
If I simply pass the given text `ctrl+alt+tab` to the kbd node's text attribute, then I'm having much difficulty getting `visit_kbd()` to differentiate between what should be a
1. \<span>+\</span>
2. \<span class="keys">...\</span>
3. \<kbd class="key-KEY_NAME">displayed text\</kbd>

I'm also trying to appease the request to consolidate the span_node and kbd_node here.

Reply:
What I had in mind is that the `kbd` role should just generate a single `kbd` node that contains the plain text without any parsing, i.e. something like: `kbd(text, text)`.  Then all of the parsing happens in `visit_kbd` --- so you would do all of the parsing and generate the full HTML all inside `visit_kbd` called on the single node generated.

Note: You can access the config from the role function via `inliner.document.settings.env.config`.  I figured that out from here: https://github.com/sphinx-doc/sphinx/blob/b4276edd848d112b4e981011c334d27cbcb20018/sphinx/util/docutils.py#L430

Note though that I don't recommend actually using SphinxRole, even though it seems like it would be the correct thing to do, because SphinxRole is badly designed (https://github.com/sphinx-doc/sphinx/pull/10251#issuecomment-1073159787).

Reply:
(But anyway if you do the parsing in `visit_kbd` I think you won't need to access the config from the role function.)

Reply:
My initial approach was to create kbd elements (nodes) from within `visit_kbd()`. I'll try it using raw HTML.

Reply:
This is satisfied now using `self.builder.config`